src/plugins/telemetry/server/event_based_telemetry/leaky_bucket.ts

```diff
@@ -0,0 +1,122 @@
+ * Side Public License, v 1.
+ */
+
+import moment from 'moment';
+import type { Logger } from 'src/core/server';
+import type { Observable, Subscription } from 'rxjs';
+import { merge, timer } from 'rxjs';
+import { throttleTime } from 'rxjs/operators';
+import type { EventEnvelope, LeakyBucketConfig, HTTPSender } from './types';
+
+export class LeakyBucket {
+  private readonly maxWaitInMs = this.config.max_wait_time.asMilliseconds();
+  private queue: Buffer[] = [];
+  private lastSend = moment();
+  private sendFailures = 0;
+  private sending = false;
+  private timer?: Subscription;
+  private onGoingGenerator?: AsyncGenerator<EventEnvelope, void>;
+  constructor(
+    private readonly logger: Logger,
+    private readonly config: LeakyBucketConfig,
+    private readonly getGenerator: () => AsyncGenerator<EventEnvelope, void>,
+    private readonly newEvents$: Observable<number>
+  ) {}
+
+  public start(sender: HTTPSender) {
+    const intervalInMs = this.config.interval.asMilliseconds();
+    this.timer = merge(timer(intervalInMs, intervalInMs), this.newEvents$)
+      .pipe(throttleTime(1000)) // Make sure we don't call this more than once per second
+      .subscribe(async () => {
+        await this.sendIfDue(sender);
+      });
+  }
+
+  public stop() {
+    this.timer?.unsubscribe();
+  }
+
+  public async sendIfDue(sender: HTTPSender) {
+    if (this.sending) return;
+
+    this.sending = true;
+    try {
+      await this.sendIfFull(sender);
+
+      this.onGoingGenerator = this.onGoingGenerator || this.getGenerator();
+      do {
+        // Retrieve elements and add them to the queue
+        const { value, done } = await this.onGoingGenerator.next();
+        if (value) {
+          const buffer = Buffer.from(JSON.stringify(value));
+          this.queue.push(buffer);
+        }
+        if (done) {
+          // Generator is consumed, clearing it out.
+          this.onGoingGenerator = void 0;
+        }
+
+        await this.sendIfFull(sender);
+      } while (this.onGoingGenerator);
+
+      if (this.hasMaxWaitExpired()) {
+        await this.send(sender);
+      }
+    } catch (err) {
+      this.logger.error(`Failed to send events: ${err.message}`);
+    } finally {
+      this.sending = false;
+    }
+  }
+
+  private async sendIfFull(sender: HTTPSender) {
+    if (this.isFullQueue()) {
+      // To don't go over our "threshold bytes per second", we need to wait 1s before consuming from queues again.
+      // Doing a Promise.all because we don't want to add the sending time to the max_frequency_of_requests wait.
+      await Promise.all([
+        this.send(sender),
+        new Promise((resolve) =>
+          setTimeout(resolve, this.config.max_frequency_of_requests.asMilliseconds())
+        ),
+      ]);
+    }
+  }
+
+  private isFullQueue(): boolean {
+    const queueSize = this.queue.reduce((acc, buffer) => acc + buffer.length, 0);
+    return queueSize >= this.config.threshold.getValueInBytes();
```

Comment:
NIT: Possible, performance improvement: precalculate `this.config.threshold.getValueInBytes()` into a private variable to avoid calling it on every loop.